internal/hostport/hostport_manager.go

```diff
@@ -121,42 +119,54 @@ func (hm *hostportManager) Add(id string, podPortMapping *PodPortMapping, natInt
 	conntrackPortsToRemove := []int{}
 	for _, pm := range hostportMappings {
 		protocol := strings.ToLower(string(pm.Protocol))
-		chain := getHostportChain(id, pm)
-		newChains = append(newChains, chain)
+		hpChain := getHostportChain(kubeHostportChainPrefix, id, pm)
+		masqChain := getHostportChain(crioMasqueradeChainPrefix, id, pm)
+		newChains = append(newChains, hpChain, masqChain)
 		if pm.Protocol == v1.ProtocolUDP {
 			conntrackPortsToRemove = append(conntrackPortsToRemove, int(pm.HostPort))
 		}
 
 		// Add new hostport chain
-		writeLine(natChains, utiliptables.MakeChainLine(chain))
+		writeLine(natChains, utiliptables.MakeChainLine(hpChain))
+		writeLine(natChains, utiliptables.MakeChainLine(masqChain))
 
-		// Prepend the new chain to KUBE-HOSTPORTS
-		// This avoids any leaking iptables rule that takes up the same port
+		// Prepend the new chains to KUBE-HOSTPORTS and CRIO-HOSTPORTS-MASQ
+		// This avoids any leaking iptables rules that take up the same port
 		writeLine(natRules, "-I", string(kubeHostportsChain),
 			"-m", "comment", "--comment", fmt.Sprintf(`"%s hostport %d"`, podFullName, pm.HostPort),
 			"-m", protocol, "-p", protocol, "--dport", fmt.Sprintf("%d", pm.HostPort),
-			"-j", string(chain),
+			"-j", string(hpChain),
 		)
-
-		// SNAT if the traffic comes from the pod itself
-		writeLine(natRules, "-A", string(chain),
+		writeLine(natRules, "-I", string(crioMasqueradeChain),
 			"-m", "comment", "--comment", fmt.Sprintf(`"%s hostport %d"`, podFullName, pm.HostPort),
-			"-s", podIP,
-			"-j", kubeMarkMasqChain)
+			"-j", string(masqChain),
+		)
 
 		// DNAT to the podIP:containerPort
 		hostPortBinding := net.JoinHostPort(podIP, strconv.Itoa(int(pm.ContainerPort)))
 		if pm.HostIP == "" || pm.HostIP == "0.0.0.0" || pm.HostIP == "::" {
-			writeLine(natRules, "-A", string(chain),
+			writeLine(natRules, "-A", string(hpChain),
 				"-m", "comment", "--comment", fmt.Sprintf(`"%s hostport %d"`, podFullName, pm.HostPort),
 				"-m", protocol, "-p", protocol,
 				"-j", "DNAT", fmt.Sprintf("--to-destination=%s", hostPortBinding))
 		} else {
-			writeLine(natRules, "-A", string(chain),
+			writeLine(natRules, "-A", string(hpChain),
 				"-m", "comment", "--comment", fmt.Sprintf(`"%s hostport %d"`, podFullName, pm.HostPort),
 				"-m", protocol, "-p", protocol, "-d", pm.HostIP,
 				"-j", "DNAT", fmt.Sprintf("--to-destination=%s", hostPortBinding))
 		}
+
+		// SNAT hairpin traffic. There is no "ctorigaddrtype" so we can't
+		// _exactly_ match only the traffic that was definitely DNATted by our
+		// rule as opposed to someone else's. But if the traffic has been DNATted
+		// and has src=dst=podIP then _someone_ needs to masquerade it, and the
+		// worst case here is just that "-j MASQUERADE" gets called twice.
```

Comment:
didn't have an issue in the past because masquerading twice? I remember you were the one that solved that issue, but I do not know if is the same case

Reply:
however, this seems very unlikely to happen

Reply:
ah, yes (https://github.com/kubernetes/kubernetes/pull/92035). But that only happened with `-j MASQUERADE --random-fully`, which we don't use here. (And was also caused by a kernel bug which has been fixed for a while now.)